### ISSUE TITLE
U-Net Denoiser

### ISSUE THREAD

**Issue by utsav-akhaury** [utsav-akhaury/ADMM#4]
The U-Net is trained on a batch of CFHT images.

When evaluated directly on input galaxy images, it denoises them and gives the desired output -
![image](https://user-images.githubusercontent.com/59925371/109665664-b51b0d00-7b6e-11eb-8510-32df6dba0ec2.png)

However, when the U-Net denoising is coupled with the FISTA subproblem within the [ADMM update steps](https://arxiv.org/pdf/1911.00443.pdf), the solution is not as expected. 

The intermediate variable on which the denoising is applied is not visually and analytically similar to the batch of images the U-Net was trained on. Furthermore, the training is also very sensitive to SNR. Our assumption is that the U-Net is overfitting to the CFHT images.

**Comment by utsav-akhaury** [comment]
The algorithm is very sensitive to the value of the augmented lagrangian parameter (ρ).
A suitable choice of ρ (obtained after some tuning) ensures very quick convergence, as shown -
![image](https://user-images.githubusercontent.com/59925371/110222112-402e3700-7ed0-11eb-9684-ed7ae43ad626.png)

**Comment by utsav-akhaury** [comment]
Better results are achieved by applying the U-Net denoising step before the FISTA update.

This is because the U-Net is practically trained to remove Additive Uncorrelated Gaussian Noise. FISTA, when applied before denoising, changes the inherent properties of the noise (for eg. it may lead to noise correlation after successive gradient descent steps).

The results still need to be investigated in detail.

**Comment by utsav-akhaury** [comment]
The value of sigma for the noise level added to the CFHT images needs to be chosen appropriately. Further investigation required.

The authors of [this ADMM Plug&Play paper](https://arxiv.org/pdf/1911.00443.pdf) train the U-Net for a range of SNR between 20 and 100.

**Comment by utsav-akhaury** [comment]
The ADMM and U-Net denoiser perform well individually. However, when the denoiser is plugged into the ADMM deconvolution code, the [solution diverges](https://github.com/utsav-akhaury/ADMM/blob/main/ADMM_tf_UNet.ipynb) after a certain no. of iterations.

To investigate the problem, I plugged in the U-Net denoiser into a simple [forward-backward algorithm](https://github.com/utsav-akhaury/ADMM/blob/main/FB_UNet.ipynb) (that uses a single gradient descent step). In this case, the solution converges quickly. 
![image](https://user-images.githubusercontent.com/59925371/112324551-3750aa00-8cb3-11eb-8b20-d0cb700103bc.png)

However, the max. intensity of the reconstructed image is around 5 times smaller than that of the target image.
![image](https://user-images.githubusercontent.com/59925371/112324623-49324d00-8cb3-11eb-9ac7-63ebeffea460.png)

Further investigation tells us that the proximal operator (in our case the U-Net denoiser) needs to be rescaled by the gradient step size.

**Comment by utsav-akhaury** [comment]
During training, when the input to the U-Net is normalised by a factor of 4000, the model seems to converge quicker. 
![image](https://user-images.githubusercontent.com/59925371/112832484-ac4a2800-9095-11eb-8e4b-ccb235f21d79.png)
However, during evaluation, certain artefacts appear in the output image. It seems that the network is overfitting to the input images and also learning the noise (which appears in the output image).
![image](https://user-images.githubusercontent.com/59925371/112832462-a48a8380-9095-11eb-96dd-5550ccffdb27.png)

**Comment by utsav-akhaury** [comment]
When Ground Truth images are fed to the U-Net iteratively, the input image changes with successive iterations (as shown [here](https://github.com/utsav-akhaury/ADMM/blob/main/UNet_test.ipynb)). Certain parts of the image are amplified. Ideally, the U-Net should only perform denoising and not affect noise-less images.

The solve this, the U-Net should be trained on a range of noise levels (sigma) instead of a constant sigma value. Also, [Spectral Normalization](https://www.tensorflow.org/addons/api_docs/python/tfa/layers/SpectralNormalization) needs to be applied to each layer of the network.

**Comment by utsav-akhaury** [comment]
Training on a range of sigma values solves the problem (for both [real](https://github.com/utsav-akhaury/ADMM/blob/main/ADMM_real_UNet.ipynb) and [parametric](https://github.com/utsav-akhaury/ADMM/blob/main/ADMM_param_UNet.ipynb) data). The ADMM + U-Net seems to converge.

Also, choosing ρ = 500 and ρ_max = 2000 (instead of ρ = 50 and ρ_max = 200) gives better results (it results in lesser residual noise around the central blob).
